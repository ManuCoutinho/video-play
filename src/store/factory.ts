import { useStore as store } from '.'
import { baseUrl } from '@/constants/baseUrl'

export async function loadInitialData() {
  store.setState({ isLoading: true })
  const controller = new AbortController()
  const signal = controller.signal
  try {
    await fetch(`${baseUrl}/courses/1`, { signal })
      .then((res) => res.json())
      .then((data) =>
        store.setState({
          course: data,
          isLoading: false
        })
      )
  } catch (err) {
    console.error(err)
    //handle error
  } finally {
    controller.abort()
  }
}

export function playNextVideo() {
  const { currentLessonIndex, currentModuleIndex, course } = store.getState()
  const nextLessonIndex = currentLessonIndex + 1
  const nextLesson =
    course?.modules[currentModuleIndex].lessons[nextLessonIndex]

  if (nextLesson) {
    store.setState({
      currentLessonIndex: nextLessonIndex
    })
  } else {
    const nextModuleIndex = currentModuleIndex + 1
    const nextModule = course?.modules[nextModuleIndex]
    if (nextModule) {
      store.setState({
        currentModuleIndex: nextModuleIndex,
        currentLessonIndex: 0
      })
    }
  }
}

export function onPlay(payload: [number, number]) {
  const [moduleIndex, lessonIndex] = payload
  store.setState({
    currentLessonIndex: lessonIndex,
    currentModuleIndex: moduleIndex
  })
}
