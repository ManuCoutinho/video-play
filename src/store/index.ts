import { create } from 'zustand'
import { PlayerStateType } from '@/models/player'
import * as factory from './factory'

type MethodsPlayer = {
  play: (payload: [number, number]) => void
  next: () => void
  load: () => Promise<void>
}
const initialState: PlayerStateType = {
  course: null,
  currentLessonIndex: 0,
  currentModuleIndex: 0,
  isLoading: true
}

const useStore = create<PlayerStateType & MethodsPlayer>(() => {
  return {
    ...initialState,
    play: (payload: [number, number]) => factory.onPlay(payload),
    next: () => factory.playNextVideo(),
    load: () => factory.loadInitialData()
  }
})

export { useStore }
