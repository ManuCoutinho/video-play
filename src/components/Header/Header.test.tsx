import '@testing-library/jest-dom'
import { describe, it } from 'vitest'
import { screen, render } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { Header } from '.'
import { state } from '@/__mocks__/data/playerState'
import { useStore as store } from '@/store'

const mockInitialState = store.getState()

describe('<Header/>', () => {
  beforeEach(() => beforeEach(() => store.setState(mockInitialState)))
  it('should render the Header component correctly', () => {
    render(<Header />)
    act(() => {
      store.setState(state)
    })
    expect(
      screen.getByRole('heading', { name: /css modules/i })
    ).toBeInTheDocument()
    expect(screen.getByText(/mÃ³dulo/i)).toBeInTheDocument()
  })
  it('should match to snapshot', () => {
    const { container } = render(<Header />)
    expect(container).toMatchSnapshot()
  })
})
