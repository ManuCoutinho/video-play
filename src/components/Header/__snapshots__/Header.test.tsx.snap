// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header/> > should match to snapshot 1`] = `
<div>
  <h1
    class="text-2xl font-semibold text-secondary"
  >
    Carregando...
  </h1>
</div>
`;
